// jest-editor-support/build/parsers/helper.js
"use strict";

require("core-js/modules/es.array.is-array.js");

require("core-js/modules/es.symbol.js");

require("core-js/modules/es.symbol.description.js");

require("core-js/modules/es.object.to-string.js");

require("core-js/modules/es.symbol.iterator.js");

require("core-js/modules/es.array.iterator.js");

require("core-js/modules/es.string.iterator.js");

require("core-js/modules/web.dom-collections.iterator.js");

require("core-js/modules/es.array.from.js");

require("core-js/modules/es.array.slice.js");

require("core-js/modules/es.function.name.js");

require("core-js/modules/es.object.define-property.js");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.parseOptions = exports.tsxPlugins = exports.tsPlugins = exports.jsPlugins = void 0;

require("core-js/modules/es.array.concat.js");

require("core-js/modules/es.regexp.exec.js");

require("core-js/modules/es.string.match.js");

function _toConsumableArray(arr) { return _arrayWithoutHoles(arr) || _iterableToArray(arr) || _unsupportedIterableToArray(arr) || _nonIterableSpread(); }

function _nonIterableSpread() { throw new TypeError("Invalid attempt to spread non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); }

function _unsupportedIterableToArray(o, minLen) { if (!o) return; if (typeof o === "string") return _arrayLikeToArray(o, minLen); var n = Object.prototype.toString.call(o).slice(8, -1); if (n === "Object" && o.constructor) n = o.constructor.name; if (n === "Map" || n === "Set") return Array.from(o); if (n === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)) return _arrayLikeToArray(o, minLen); }

function _iterableToArray(iter) { if (typeof Symbol !== "undefined" && iter[Symbol.iterator] != null || iter["@@iterator"] != null) return Array.from(iter); }

function _arrayWithoutHoles(arr) { if (Array.isArray(arr)) return _arrayLikeToArray(arr); }

function _arrayLikeToArray(arr, len) { if (len == null || len > arr.length) len = arr.length; for (var i = 0, arr2 = new Array(len); i < len; i++) { arr2[i] = arr[i]; } return arr2; }

/**
 * Copyright (c) 2014-present, Facebook, Inc. All rights reserved.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 */
var commonPlugins = ['asyncGenerators', 'bigInt', 'classPrivateMethods', 'classPrivateProperties', 'classProperties', 'doExpressions', 'dynamicImport', 'estree', 'exportDefaultFrom', 'exportNamespaceFrom', // deprecated
'functionBind', 'functionSent', 'importMeta', 'logicalAssignment', 'nullishCoalescingOperator', 'numericSeparator', 'objectRestSpread', 'optionalCatchBinding', 'optionalChaining', 'partialApplication', 'throwExpressions', 'topLevelAwait', ['decorators', {
  decoratorsBeforeExport: true
}], ['pipelineOperator', {
  proposal: 'smart'
}]];
var jsPlugins = [].concat(commonPlugins, ['flow', 'jsx']);
exports.jsPlugins = jsPlugins;
var tsPlugins = [].concat(commonPlugins, ['typescript']);
exports.tsPlugins = tsPlugins;
var tsxPlugins = [].concat(commonPlugins, ['typescript', 'jsx']);
exports.tsxPlugins = tsxPlugins;

var parseOptions = function parseOptions(filePath) {
  var strictMode = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : false;

  if (filePath.match(/\.ts$/i)) {
    return {
      plugins: _toConsumableArray(tsPlugins)
    };
  }

  if (filePath.match(/\.tsx$/i)) {
    return {
      plugins: _toConsumableArray(tsxPlugins)
    };
  } // for backward compatibility, use js parser as default unless in strict mode


  if (!strictMode || filePath.match(/\.m?jsx?$/i)) {
    return {
      plugins: _toConsumableArray(jsPlugins)
    };
  }

  throw new TypeError("unable to find parser options for unrecognized file extension: ".concat(filePath));
};

exports.parseOptions = parseOptions;
